File: Src/FUSE/Linux/x86-64/I-O_Engine/confirm.c
/*
 * confirm.c — cafs_confirm() and cafs_close().
 *
 * confirm() writes, in order, each followed by fdatasync(): (1) the
 * primary Superblock with last_mount_time refreshed and its BLAKE3-128
 * checksum recomputed; (2) the same content to Superblock Backup (via
 * LBA0's fallback pointer), if one exists; (3) the recomputed primary
 * Parity Block, XOR(new Superblock, cached Config Snapshot); (4) if a
 * usable M-SMART location was found at mount(), mount_count++ and
 * clean_unmount_flag=0 on it directly. Only the primary Superblock
 * write (1) is fatal to the call; (2)-(4) failing is logged and
 * non-fatal, consistent with "ignore all anchor block errors except
 * Superblock" for this PoC. There is no Backup Parity Block to keep
 * in sync — the revised design (see cafs_io_layout.h) drops
 * end-of-device redundancy for everything except Superblock and
 * SMART, Parity included.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2026, Imran Bin Gifary (System Delta)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "cafs_io_internal.h"
#include "syscalls.h"

cafs_status_t cafs_confirm(cafs_ctx_t *ctx)
{
    if (!ctx) return CAFS_ERR_INVALID_ARG;
    if (!ctx->anchor_loaded) return CAFS_ERR_NOT_MOUNTED;
    if (ctx->opts.readonly || !ctx->want_write) return CAFS_ERR_READONLY;

    pthread_mutex_lock(&ctx->lock);

    int64_t now = (int64_t)time(NULL);
    memcpy(ctx->superblock + SB_OFF_LAST_MOUNT_TIME, &now, 8);

    uint8_t sb_checksum[16];
    cafs_blake3_128(ctx->superblock, SB_HASHED_LEN, sb_checksum);
    memcpy(ctx->superblock + SB_OFF_CHECKSUM, sb_checksum, 16);

    cafs_status_t st = cafs_write_raw(ctx, ctx->lba0.real_lba, CAFS_REAL_BLOCK_4K, ctx->superblock);
    if (st != CAFS_OK) {
        pthread_mutex_unlock(&ctx->lock);
        return st;
    }
    cafs_sys_fdatasync(ctx->fd);

    if (ctx->lba0.fallback_real_lba != 0) {
        if (cafs_write_raw(ctx, ctx->lba0.fallback_real_lba, CAFS_REAL_BLOCK_4K, ctx->superblock) == CAFS_OK) {
            cafs_sys_fdatasync(ctx->fd);
            memcpy(ctx->superblock_backup, ctx->superblock, CAFS_REAL_BLOCK_4K);
            ctx->superblock_backup_valid = 1;
        } else {
            fprintf(stderr, "cafs: confirm: primary Superblock updated but the backup "
                             "write failed (non-fatal, primary is authoritative)\n");
        }
    }

    if (ctx->lba2.real_lba != 0) {
        uint8_t new_parity[CAFS_REAL_BLOCK_4K];
        for (size_t i = 0; i < CAFS_REAL_BLOCK_4K; i++)
            new_parity[i] = ctx->superblock[i] ^ ctx->config_snapshot[i];
        if (cafs_write_raw(ctx, ctx->lba2.real_lba, CAFS_REAL_BLOCK_4K, new_parity) == CAFS_OK) {
            cafs_sys_fdatasync(ctx->fd);
            memcpy(ctx->parity, new_parity, CAFS_REAL_BLOCK_4K);
        } else {
            fprintf(stderr, "cafs: confirm: Parity Block recompute write failed (non-fatal)\n");
        }
    }

    if (ctx->smart_main_lba != 0) {
        uint8_t smart_buf[CAFS_REAL_BLOCK_4K];
        memcpy(smart_buf, ctx->smart_main, CAFS_REAL_BLOCK_4K);

        uint32_t magic;
        memcpy(&magic, smart_buf + SMART_OFF_MAGIC, 4);

        uint32_t mount_count;
        if (magic == SMART_MAGIC) {
            memcpy(&mount_count, smart_buf + SMART_OFF_MOUNT_COUNT, 4);
            mount_count++;
        } else {
            /* Uninitialized/garbage M-SMART slot: start fresh rather
             * than fail — matches "won't crash" and this PoC's
             * best-effort treatment of everything but Superblock. */
            memset(smart_buf, 0, sizeof(smart_buf));
            uint32_t magic_val = SMART_MAGIC;
            memcpy(smart_buf + SMART_OFF_MAGIC, &magic_val, 4);
            uint32_t ver = 1;
            memcpy(smart_buf + SMART_OFF_VERSION, &ver, 4);
            uint64_t seq = 1;
            memcpy(smart_buf + SMART_OFF_SEQUENCE, &seq, 8);
            mount_count = 1;
        }
        memcpy(smart_buf + SMART_OFF_MOUNT_COUNT, &mount_count, 4);
        smart_buf[SMART_OFF_CLEAN_UNMOUNT_FLAG] = 0;

        uint8_t smart_checksum[16];
        cafs_blake3_128(smart_buf, SMART_HASHED_LEN, smart_checksum);
        memcpy(smart_buf + SMART_OFF_CHECKSUM, smart_checksum, 16);

        if (cafs_write_raw(ctx, ctx->smart_main_lba, CAFS_REAL_BLOCK_4K, smart_buf) == CAFS_OK) {
            cafs_sys_fdatasync(ctx->fd);
            memcpy(ctx->smart_main, smart_buf, CAFS_REAL_BLOCK_4K);
            ctx->smart_main_valid = 1;
        } else {
            fprintf(stderr, "cafs: confirm: M-SMART update write failed (non-fatal)\n");
        }
    }

    pthread_mutex_unlock(&ctx->lock);
    return CAFS_OK;
}

/* Not part of the originally-named four-stage flow — see cafs_io.h's
 * doc comment: without a counterpart, destroy_flush and a graceful fd
 * close never happen. */
cafs_status_t cafs_close(cafs_ctx_t *ctx)
{
    if (!ctx) return CAFS_ERR_INVALID_ARG;
    if (ctx->is_open) {
        if (ctx->opts.destroy_flush && ctx->want_write)
            cafs_sys_fdatasync(ctx->fd);
        cafs_sys_close(ctx->fd);
        ctx->is_open = 0;
    }
    pthread_mutex_destroy(&ctx->lock);
    free(ctx);
    return CAFS_OK;
}

File: Src/FUSE/Linux/x86-64/I-O_Engine/src/confirm.rs
//! `CafsCtx::confirm` — sets Superblock `last_mount_time = now`,
//! recomputes its BLAKE3-128 checksum, and commits it via fs.info
//! §13.3 (In-Place Content Update): primary Superblock, then
//! Superblock Backup (via LBA0's fallback), then the primary Parity
//! Block (XOR of the new Superblock and the cached Config Snapshot).
//! Each write is followed by `fdatasync()`. There is no Backup Parity
//! Block to keep in sync — Parity has no backup under the revised
//! redundancy design (ADR-006).
//!
//! Also increments `mount_count` and sets `clean_unmount_flag = 0` on
//! M-SMART directly, best-effort (ADR-008) — a missing/unavailable
//! SMART table doesn't fail the call. Only the primary Superblock
//! write is fatal to `confirm()`.
//!
//! SPDX-License-Identifier: BSD-3-Clause
//! Copyright (c) 2026, Imran Bin Gifary (System Delta)

use std::time::{SystemTime, UNIX_EPOCH};

use crate::block_io::write_raw;
use crate::checksum;
use crate::error::{CafsError, CafsResult};
use crate::layout::*;
use crate::CafsCtx;

fn now_unix() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

impl CafsCtx {
    pub fn confirm(&mut self) -> CafsResult<()> {
        if self.opts.readonly || !self.want_write {
            return Err(CafsError::Readonly);
        }

        // 1. Primary Superblock: last_mount_time = now, recompute
        // checksum, write, fsync.
        let now = now_unix();
        self.superblock[SB_OFF_LAST_MOUNT_TIME..SB_OFF_LAST_MOUNT_TIME + 8]
            .copy_from_slice(&now.to_le_bytes());
        let cksum = checksum::blake3_128(&self.superblock[0..SB_HASHED_LEN]);
        self.superblock[SB_OFF_CHECKSUM..SB_OFF_CHECKSUM + SB_CHECKSUM_LEN].copy_from_slice(&cksum);

        write_raw(self, self.lba0.real_offset, &self.superblock.clone())?;
        rustix::fs::fdatasync(&self.fd)?;

        // 2. Superblock Backup, same content — best-effort, logged
        // and non-fatal on failure (ADR-007/008 posture: only the
        // primary write is load-bearing for the call to succeed).
        if let Some((backup_offset, _)) = self.lba0.fallback {
            if let Err(e) = write_raw(self, backup_offset, &self.superblock.clone()) {
                eprintln!("cafs: confirm(): failed to update Superblock Backup at offset {}: {e}", backup_offset.0);
            } else if let Err(e) = rustix::fs::fdatasync(&self.fd) {
                eprintln!("cafs: confirm(): fdatasync after Superblock Backup write failed: {e}");
            }
        }

        // 2.5. Primary Parity Block = XOR(new Superblock, cached
        // Config Snapshot) — only possible if Config Snapshot was
        // successfully read at mount() (best-effort, ADR-007).
        if let Some(cfg) = &self.config_snapshot {
            let mut parity = [0u8; REAL_BLOCK_4K as usize];
            for i in 0..REAL_BLOCK_4K as usize {
                parity[i] = self.superblock[i] ^ cfg[i];
            }
            if let Err(e) = write_raw(self, self.parity_offset, &parity) {
                eprintln!("cafs: confirm(): failed to update Parity Block at offset {}: {e}", self.parity_offset.0);
            } else if let Err(e) = rustix::fs::fdatasync(&self.fd) {
                eprintln!("cafs: confirm(): fdatasync after Parity Block write failed: {e}");
            }
        }

        // M-SMART: best-effort, never fatal.
        if let Some((mut smart, offset)) = self.smart_main {
            let mount_count = u32::from_le_bytes(
                smart[SMART_OFF_MOUNT_COUNT..SMART_OFF_MOUNT_COUNT + 4].try_into().unwrap(),
            );
            smart[SMART_OFF_MOUNT_COUNT..SMART_OFF_MOUNT_COUNT + 4]
                .copy_from_slice(&mount_count.wrapping_add(1).to_le_bytes());
            smart[SMART_OFF_CLEAN_UNMOUNT_FLAG] = 0;
            let sc = checksum::blake3_128(&smart[0..SMART_HASHED_LEN]);
            smart[SMART_OFF_CHECKSUM..SMART_OFF_CHECKSUM + SMART_CHECKSUM_LEN].copy_from_slice(&sc);

            if let Err(e) = write_raw(self, offset, &smart) {
                eprintln!("cafs: confirm(): failed to update M-SMART at offset {}: {e}", offset.0);
            } else {
                if let Err(e) = rustix::fs::fdatasync(&self.fd) {
                    eprintln!("cafs: confirm(): fdatasync after M-SMART write failed: {e}");
                }
                self.smart_main = Some((smart, offset));
            }
        }

        Ok(())
    }
}
